internal/statesync/reactor.go

```diff
@@ -379,6 +391,12 @@ func (r *Reactor) Sync(ctx context.Context) (sm.State, error) {
 		}
 	}
 
+	if r.postSyncHook != nil {
+		if err := r.postSyncHook(ctx, state); err != nil {
```

Comment:
I'd mildly prefer just giving the reactor a clear reference to the consensus reactor so it's easier to trace your way through when debugging. I'm not sure much is really gained by adding an additional 'hook' callback field on the reactor.

The style I somewhat like for this is defining an interface within the package that's consuming the outside code. Here, the interface would be:

```go
type consensusReactor interface {
        SwitchToConsensus(ctx context.Context, state sm.State, skipWAL bool)
}
```
The consuming code can then accept that interface and invoke it when required. Invocations that don't need to actually perform the switch can just take a nop version or a mock.

It looks like this is effectively exactly what the blocksync reactor is doing:

https://github.com/tendermint/tendermint/blob/2304ea70f7f98cf1759510b24e3a95ddecb45d2a/internal/blocksync/reactor.go#L56

Reply:
I'm not sure If I like this solution either. It kind of tangles the consensus reactor together. Ultimately I'd prefer if the node dictated this i.e. it ran a method on the state sync reactor to sync state and once that terminated it would then start up the consensus reactor. This way reactors don't need to be aware of each other rather focus on serving the function they're meant for

Reply:
This pattern is how the blocksync reactor coordinates with consensus, and I've never much liked it, to be honest, though it does work. 

I think ultimately that better than either of these approaches (i.e. yours or mine as implemented,) we should:
- reactors should identify if they're syncing or not syincing and then have some sort of coordinating "metaservice" that switches between them so that reactors don't need to know about each other at all. 
- reactors should be able to send messages to each other that say "hey turn on," or "hey turn off". 

The first seems like a thing that we could do (and in fact has been the prevailing design that we've been talking about this for a while, but I opted to avoid because it's a big undertaking and I've been trying to move us out of the current situation in a more iterative manner because this is all tested in the e2e tests, and I want to keep the steps small to avoid breaking things too much before the release. I also sort of feel that while this would be a lot of code (and a lot of new code), it just sort of moves the pieces around on our plate without actually providing all that much benefit,) 

The second option has always appealed to me, but feels like a thing we should do after the libp2p stuff lands and we've cut out a lot of the now-legacy p2p code (so that we can have local messages that can easily bypass the network stack and the serialization requirements). 

In light of this, the hook kind of appeals to me, because it means that the reactors don't need to know about each other at all, and we don't need to mock out any bit of the reactor for testing. I also think it'll be easy to rip out and unwind later, which I think is still something I'd like to see happen.

Reply:
I guess we do have a mechanism for reactors to communicate via the event bus, which could permit signaling like "blocksync cannot make progress, going to sleep", which would permit (say) the consensus reactor to "wake up", and likewise when the node starts up, consensus might stall until it sees that event go by.

On that basis, I think I'm mostly in agreement with you about the second option, both on the basis that it's a cleaner design and that it's more easily reachable from where we are now without deepening some of the strange dependency clustering we inherited.

Reply:
> I'm not sure If I like this solution either. It kind of tangles the consensus reactor together. Ultimately I'd prefer if the node dictated this i.e. it ran a method on the state sync reactor to sync state and once that terminated it would then start up the consensus reactor. This way reactors don't need to be aware of each other rather focus on serving the function they're meant for

I mean I'd argue that injection the function here entangles them *less* than one reactor needed to hold a reference to another reactor, (again, this relative to the way blocksync does things).

Reply:
> I'm not sure If I like this solution either.
I was referring to Williams alternative

Reply:
> reactors should identify if they're syncing or not syincing and then have some sort of coordinating "metaservice" that switches between them so that reactors don't need to know about each other at all.

Yeah, I'd prefer this as well. 

The pattern here feels reminiscent of the way that some of the methods are implemented in the consensus reactor, passing hook functions that are defined elsewhere. It's frustrating to debug because you have to jump around to make sure you're looking at the right implementation from the particular case so I'd mildly prefer we just be explicit about what we're wiring together.

Reply:
I'm inclined to let this go and continue to push on future iterations/unpacking the relationships and interactions of these components in future work, particularly if we're agreed that: 

- this implementation is a little better than the current state of the code base 
- it's important to move slowly on this package because even small changes have subtle ripple effects. 
- the ideal designs for this aspect of the code aren't achievable in the range of 0.36.

> I guess we do have a mechanism for reactors to communicate via the event bus, 

While I am terrified by this prospect, it might work. I think I'm still partial to doing this kind of communication via the p2p layer instead, because being able to think about reactors as more distinct services (and while I'm not super eager to push things into different processes, I think it's useful for vague multi-tendermint/heterogenous consensus reasons).